text/0022-pibd-deployment.md

```diff
@@ -0,0 +1,444 @@
+		* If it is a left sibling, the next element will be the corresponding right sibling, so no further action is needed.
+		* If it is a right sibling, calculate and append hashes into the MMR up to the next leaf insertion index
+	* If the next element to be inserted is a pruned subtree (i.e. a hash value,) add the subtree to the MMR then calculate and append hashes up to the position of the next leaf insertion index. These hash values that should also be supplied within the segments's list of hashes, so the node should ensure the calculated hashes match the segment's data.
+
+#### Example of appending segment data to an MMR
+
+Assume the current state of the MMR is as follows:
+
+```
+Height
+1         2    
+         / \   
+0       0   1 
+
+Size: 3
+Last Position: 2
+```
+
+The next segment element in the sorted list is a hash (a pruned subtree) at position 5. We append this hash to the MMR, (both leaves underneath it have been pruned)
+
+```
+Height
+1         2      5  
+         / \    / \ 
+0       0   1 (3) (4)
+
+Size: 6
+Last Position: 5
+```
+
+Then calculate hashes up to the position of `next leaf index - 1` (i.e. 7 - 1):
+
+```
+Height
+2             6
+           /     \
+1         2       5  
+         / \     / \ 
+0       0   1  (3) (4)
+
+Size: 7
+Last Position: 6
+```
+
+The next segment element will be the hash at 6, we check this matches our calculated hash and continue.
+
+At this point, the next elements could be (for example):
+
+* Leaves at 7 and 8 - in which case we'd simply apply the leaves and calculate hashes to the next leaf insertion index, giving:
+
+```
+Height
+2             6       
+           /     \    
+1         2       5       9 
+         / \     / \     / \ 
+0       0   1  (3) (4)  7   8
+
+Size: 10
+Last Position: 9
+```
+
+* A hash for 9 (if leaves 7 and 8 are pruned), in which case we'd apply the pruned subtree and calculate hashes to the next insertion index, giving:
+
+```
+Height
+2             6       
+           /     \    
+1         2       5       9 
+         / \     / \     / \ 
+0       0   1  (3) (4) (7) (8)
+
+Size: 10
+Last Position: 9
+```
+
+* A hash at 13 (if leaves 7, 8, 10 and 11 are pruned), in which case we apply the pruned subtree and calculate hashes to the next insertion index, giving:
+
+```
+Height
+3                     14 
+                /             \
+               /               \
+2             6                 13   
+           /     \          /        \
+1         2       5       (9)       (12)
+         / \     / \      / \      /    \ 
+0       0   1  (3) (4)  (7) (8)  (10)  (11)
+
+Size: 14
+Last Position: 13
+```
+
+This process continues until all hash and leaf values in the segment have been exhausted.
+
+If there is a mismatch anywhere or an error resulting from applying an element to an MMR, the node should wipe all data and restart the PIBD process. (Note this would most likely be the result of an internal or implementation error.)
```

Comment:
Before wiping all data, we should also log any information that could help debug this unexpected event.

Reply:
Added the note